Exercise/src/my_string.cpp

```diff
@@ -173,29 +173,229 @@ Exercise_1::CMyString& Exercise_1::CMyString::operator=(const Exercise_1::CMyStr
+		return substring;
+	}
+
+	const size_t cnEndPosition = nPosition + nCharactersCount;
+	for(size_t nCurrentIndex = nPosition; nCurrentIndex < cnEndPosition; nCurrentIndex++)
+	{
+		substring.m_szData[nCurrentIndex - nPosition] = m_szData[nCurrentIndex];
+	}
+
+	return substring;
+}
+std::optional<size_t> MyStructs::CMyString::Find(const char *cszStringToFind) const noexcept
+{
+	if(cszStringToFind == nullptr)
+	{
+		return std::nullopt;
+	}
+	
+	size_t nStringLength = GetStringLength(cszStringToFind);
+	if(nStringLength <= 1)
+	{
+		return std::nullopt;
+	}
+	nStringLength -= 1; // Excluding terminator
+
+	size_t nIndex = 0; // Index for iterating through this.size()
+	size_t nCurrentStringIndex = 0; // Index for iterating through cszStringToFind
+	bool bIsLookingFor = false;
+
+	for(; nIndex < m_nSize; nIndex++)
+	{
+		if(m_szData[nIndex] != cszStringToFind[nCurrentStringIndex])
+		{
+			nCurrentStringIndex = 0;
+			bIsLookingFor = false;
+		} else if(m_szData[nIndex] == cszStringToFind[nCurrentStringIndex]){
+			++nCurrentStringIndex;
+			bIsLookingFor = true;
+			// Breaking if we reached end of this string
+			if(nCurrentStringIndex == nStringLength)
+			{
+				break;
+			}
+
+		}
+	}
+
+	return bIsLookingFor ? std::optional<size_t>(nIndex - nStringLength + 1) : std::nullopt;
+}
+
+
+MyStructs::CMyString MyStructs::CMyString::Trim(size_t nPosition, size_t nCharactersCount) noexcept
+{
+	CMyString newString{nullptr};
+	if(IsOutOfBounds(nPosition, nCharactersCount))
+	{
+		return newString;
+	}
+	
+	// Checking if we should just swap values
+	if(nCharactersCount == size() - 1) 
+	{
+		std::swap(m_nSize, newString.m_nSize);
+		std::swap(m_szData, newString.m_szData);
+		return newString;
+	}
+	
+	// Checking if we are triming terminator
+	const size_t cnOldSize = size();
+	char *cszOldString = m_szData;
+
+	const bool cbIsReachedTerminator = IsReachedTerminator(nPosition, nCharactersCount);
+	m_nSize = cnOldSize - nCharactersCount;
+	newString.m_nSize = nCharactersCount + 1;
+
+	if(cbIsReachedTerminator)
+	{
+		++m_nSize;
+		--newString.m_nSize; // If we already will copy terminator reducing size by 1
+	}
+
+	newString.TryToAllocate(newString.m_nSize);
+	TryToAllocate(m_nSize);
+	// If we cannot allocate new data, backing up to previous and exiting
+	if(!newString.data() || !data())
+	{
+		m_szData = cszOldString;
+		m_nSize = cnOldSize;
+		return newString;
+	}
+
+	const size_t endPosition = nPosition + nCharactersCount;
+	for(size_t nIndex = 0; nIndex < cnOldSize - 1; nIndex++){
+		if(nIndex < nPosition)
+		{
+			m_szData[nIndex] = cszOldString[nIndex];
+		} else if(nIndex >= endPosition)
+		{
+			m_szData[nIndex - newString.size() + 1] = cszOldString[nIndex];
+		} else 
+		{
+			newString.m_szData[nIndex - nPosition] = cszOldString[nIndex];
+		}
+		
+	}
+
+	// Manually setting terminators in case we have deleted them
+	m_szData[size() - 1] = '\0';
+	newString[newString.size() - 1] = '\0';
+
+	delete[] cszOldString;
+	return newString;
+}
```

Comment:
Rewrite this func according to our discussion

Reply:
Now this function trims all characters that match filter